### ISSUE TITLE
 no implicit conversion of nil into String (TypeError)

### ISSUE THREAD

**Issue by bronzdoc** [how-is/how_is#222]
Hi :wave:

I just tried to use the `how_is` gem but couldn't run one of the examples in the README, i'm doing something wrong?

```shell
$ how_is rubygems/rubygems 2018-03-01 --output report-2018-03-01.html

Traceback (most recent call last):
        20: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `<main>'
        19: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `eval'
        how_is rubygems/rubygems 2018-03-01 --output report-2018-03-01.html
Traceback (most recent call last):
        20: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `<main>'
        19: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `eval'
        18: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `<main>'
        17: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `load'
        16: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/exe/how_is:5:in `<top (required)>'
        15: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        14: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        13: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is.rb:4:in `<top (required)>'
        12: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        11: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        10: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/report.rb:4:in `<top (required)>'
         9: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         8: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `requ how_is rubygems/rubygems 2018-03-01 --output report-2018-03-01.html
Traceback (most recent call last):
        20: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `<main>'
        19: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `eval'
        18: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `<main>'
        17: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `load'
        16: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/exe/how_is:5:in `<top (required)>'
        15: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        14: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        13: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is.rb:4:in `<top (required)>'
        12: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        11: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        10: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/report.rb:4:in `<top (required)>'
         9: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         8: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         7: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github/contributions.rb:4:in `<top (required)>'
         6: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         5: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         4: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:8:in `<top (required)>'
         3: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:9:in `<module:HowIs>'
         2: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:10:in `<module:Sources>'
         1: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `<class:Github>'
/home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `+': no implicit conversion of nil into String (TypeError)
ire'
         7: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github/contributions.rb:4:in `<top (required)>'
         6: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         5: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         4: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:8:in `<top (required)>'
         3: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:9:in `<module:HowIs>'
         2: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:10:in `<module:Sources>'
         1: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `<class:Github>'
/home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `+': no implicit conversion of nil into String (TypeError)
 18: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `<main>'
        17: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `load'
        16: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/exe/how_is:5:in `<top (required)>'
        15: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        14: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        13: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is.rb:4:in `<top (required)>'
        12: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        11: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        10: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/report.rb:4:in `<top (required)>'
         9: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         8: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         7: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github/contributions.rb:4:in `<top (required)>'
         6: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         5: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `re how_is rubygems/rubygems 2018-03-01 --output report-2018-03-01.html
Traceback (most recent call last):
        20: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `<main>'
        19: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/ruby_executable_hooks:15:in `eval'
        18: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `<main>'
        17: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/bin/how_is:23:in `load'
        16: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/exe/how_is:5:in `<top (required)>'
        15: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        14: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        13: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is.rb:4:in `<top (required)>'
        12: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        11: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        10: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/report.rb:4:in `<top (required)>'
         9: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         8: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         7: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github/contributions.rb:4:in `<top (required)>'
         6: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         5: from /usr/share/rvm/rubies/ruby-2.5.0/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         4: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:8:in `<top (required)>'
         3: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:9:in `<module:HowIs>'
         2: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:10:in `<module:Sources>'
         1: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `<class:Github>'
/home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `+': no implicit conversion of nil into String (TypeError)
quire'
         4: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:8:in `<top (required)>'
         3: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:9:in `<module:HowIs>'
         2: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:10:in `<module:Sources>'
         1: from /home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `<class:Github>'
/home/bronzdoc/.rvm/gems/ruby-2.5.0@how_is/gems/how_is-23.0.0/lib/how_is/sources/github.rb:17:in `+': no implicit conversion of nil into String (TypeError)
```

**Comment by duckinator** [comment]
**WORKAROUND:** use Ruby 2.4.x.

Thanks for reporting this! I can reproduce with 2.5.0, but not on 2.3.x or 2.4.x. I've added 2.5.0 to the test matrix, and am working from there.

**Comment by bronzdoc** [comment]
I tried with with 2.4.x but still getting the same error

**Comment by bronzdoc** [comment]
Oh i see what is the problem:
```
BASIC_AUTH    = ENV["HOWIS_BASIC_AUTH"]
ACCESS_TOKEN  = ENV["HOWIS_GITHUB_TOKEN"]
```
  I need those two environment variables to make it work properly, maybe add that to the README?

**Comment by duckinator** [comment]
oh! okay, so the failure i got on 2.5.0 was a bit misleading &mdash; that was also on a system without those variables defined.

I'll document those, and how to create them, then close this issue once that's done. :+1:

**Comment by duckinator** [comment]
I'll also make the error actually remotely useful lmao

**Comment by bronzdoc** [comment]
:joy: thanks @duckinator !

**Comment by duckinator** [comment]
1. [Updated the README.](https://github.com/how-is/how_is/blob/master/README.md#configuration)
2. [Released v24.0.0](https://github.com/how-is/how_is/blob/master/CHANGELOG.md#v2400-2018-03-17) with better error messages if the environment variables are undefined.
